runtime-sdk/modules/evm/src/signed_query.rs

```diff
@@ -0,0 +1,248 @@
+    }
+
+    let block_hashes = state::block_hashes(state);
+    if let Some(hash) = block_hashes.get::<_, Hash>(&leash.block_number.to_be_bytes()) {
+        if hash.as_ref() != leash.block_hash.as_ref() {
+            return Err(Error::InvalidSignedQuery("unexpected base block"));
+        }
+        let block_delta = current_block
+            .checked_sub(leash.block_number)
+            .unwrap_or_else(|| leash.block_number - current_block);
+        if block_delta > leash.block_range {
+            return Err(Error::InvalidSignedQuery("current block out of range"));
+        }
+    } else {
+        return Err(Error::InvalidSignedQuery("base block not found"));
+    }
+
+    Ok(query)
+}
+
+macro_rules! leash_type_str {
+    () => {
+        concat!(
+            "Leash",
+            "(",
+            "uint64 nonce",
+            ",uint64 blockNumber",
+            ",uint256 blockHash",
+            ",uint64 blockRange",
+            ")",
+        )
+    };
+}
+
+fn hash_call_toplevel<Cfg: Config>(query: &SimulateCallQuery) -> [u8; 32] {
+    let call_struct_hash = hash_call(query);
+    let domain_separator = hash_domain::<Cfg>();
+    let mut encoded_call = [0u8; 66];
+    encoded_call[0..2].copy_from_slice(b"\x19\x01");
+    encoded_call[2..34].copy_from_slice(domain_separator);
+    encoded_call[34..].copy_from_slice(&call_struct_hash);
+    Keccak256::digest(&encoded_call).into()
+}
+
+fn hash_call(query: &SimulateCallQuery) -> [u8; 32] {
+    const CALL_TYPE_STR: &str = concat!(
+        "Call",
+        "(",
+        "address from",
+        ",address to",
+        ",uint256 value",
+        ",uint256 gasPrice",
+        ",uint64 gasLimit",
+        ",bytes data",
+        ",Leash leash",
+        ")",
+        leash_type_str!()
+    );
+    hash_encoded(&[
+        encode_bytes(CALL_TYPE_STR),
+        Token::Address(query.caller.0.into()),
+        Token::Address(query.address.0.into()),
+        Token::Uint(ethabi::ethereum_types::U256(query.value.0)),
+        Token::Uint(ethabi::ethereum_types::U256(query.gas_price.0)),
+        Token::Uint(query.gas_limit.into()),
+        encode_bytes(&query.data),
+        Token::Uint(hash_leash(query.leash.as_ref().unwrap() /* checked above */).into()),
+    ])
+}
+
+fn hash_leash(leash: &Leash) -> [u8; 32] {
+    hash_encoded(&[
+        encode_bytes(leash_type_str!()),
+        Token::Uint(leash.nonce.into()),
+        Token::Uint(leash.block_number.into()),
+        Token::Uint(leash.block_hash.0.into()),
+        Token::Uint(leash.block_range.into()),
+    ])
+}
+
+fn hash_domain<Cfg: Config>() -> &'static [u8; 32] {
+    static DOMAIN_SEPARATOR: OnceCell<[u8; 32]> = OnceCell::new(); // Not `Lazy` because of generic.
+    DOMAIN_SEPARATOR.get_or_init(|| {
+        const DOMAIN_TYPE_STR: &str = "EIP712Domain(string name,string version,uint256 chainId)";
+        hash_encoded(&[
+            encode_bytes(DOMAIN_TYPE_STR),
+            encode_bytes("Sapphire ParaTime"),
```

Comment:
This should be generic and possibly follow our domain separation context scheme? E.g.
```
oasis-runtime-sdk/evm: signed query
```